scipy/optimize/tests/test_optimize.py

```diff
@@ -2239,6 +2241,198 @@ def test_gh12696():
+                assert len(x) == 2
+                return 1
+
+            constr = optimize.NonlinearConstraint(con, 2, 3)
+            res = optimize.minimize(
+                optimize.rosen, x0, method=method, bounds=bounds,
+                constraints=[constr]
+            )
+            assert res.success
+            assert_allclose(res.x, np.r_[constr_best_x.x, 2.0], rtol=3e-6)
+
+            constr = optimize.NonlinearConstraint(con, 2, 3, jac=conjac)
+            res = optimize.minimize(
+                optimize.rosen, x0, method=method, bounds=bounds,
+                constraints=[constr]
+            )
+            assert res.success
+            assert_allclose(res.x, np.r_[constr_best_x.x, 2.0], rtol=3e-6)
+
+        # check that minimize works if func returns func and grad.
+        if method not in ['Powell']:
+            # powell doesn't use jac
+            res = optimize.minimize(
+                func_and_grad, x0, method=method, bounds=bounds, jac=True
+            )
+            assert res.success
+            assert_allclose(res.x, np.r_[best_x.x, 2.0], rtol=3e-6)
+
+# --- Test minimize with equal upper and lower bounds --- #
+
+np.random.seed(0)
+eb_x0 = np.random.rand(4)
+eb_lb = np.array([0, 2, -1, -1])
+eb_ub = np.array([3, 2, 2, -1])
+eb_i = (eb_lb == eb_ub)
+
+
+def eb_rosen(x):
+    # assert x.size == 4
+    # assert_allclose(x[eb_i], eb_lb[eb_i])
+    return optimize.rosen(x)
+
+
+def eb_drosen(x):
+    # assert x.size == 4
+    # assert_allclose(x[eb_i], eb_lb[eb_i])
+    return optimize.rosen_der(x)
+
+
+def eb_callback(x, *args):
+    assert x.size == 4
+    assert_allclose(x[eb_i], eb_lb[eb_i])
+
+
+# test without gradient, with gradient, and with combined objective/gradient
+eb_kwds = ({"fun": eb_rosen, "jac": False},
+           {"fun": eb_rosen, "jac": eb_drosen},
+           {"fun": (lambda x: (eb_rosen(x), eb_drosen(x))),
+            "jac": True})
+
+# test with both old- and new-style bounds
+eb_bound_types = (lambda lb, ub: list(zip(lb, ub)), Bounds)
+
+
+# constraints and jacobians
+def eb_constraint1(x):
+    return x[0:1] - 1
+
+
+def eb_dconstraint1(x):
+    dc = np.zeros_like(x)
+    dc[0] = 1
+    return dc
+
+
+def eb_constraint2(x):
+    return x[2:3] - 0.5
+
+
+def eb_dconstraint2(x):
+    dc = np.zeros_like(x)
+    dc[2] = 1
+    return dc
+
+
+c1a = NonlinearConstraint(eb_constraint1, -np.inf, 0)
+c1b = NonlinearConstraint(eb_constraint1, -np.inf, 0, eb_dconstraint1)
+c2a = NonlinearConstraint(eb_constraint2, -np.inf, 0)
+c2b = NonlinearConstraint(eb_constraint2, -np.inf, 0, eb_dconstraint2)
+
+# Test for many combinations of constraints w/ and w/out jacobian
+# Pairs in format: (test constraints, reference constraints)
+# (always use analytical jacobian in reference)
+eb_constraints = (([], []), (c1a, c1b), (c2b, c2b), ([c1b], [c1b]),
+                  ([c2a], [c2b]), ([c1a, c2a], [c1b, c2b]),
+                  ([c1a, c2b], [c1b, c2b]), ([c1b, c2b], [c1b, c2b]))
```

Comment:
The idea is to test:
* no constraints
* single constraint without analytical jacobian
* single constraint with analytical jacobian
* list with a single constraint with analytical jacobian
* list with a single constraint without analytical jacobian
* lists with multiple constraints, all without analytical jacobian
* lists with multiple constraints, some with analytical jacobian
* lists with multiple constraints, all with analytical jacobian

with all the possibilities for the function/gradient specification:
* no analytical gradient
* analytical gradient defined separately
* analytical gradient returned with function value